infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationAckReader.java

```diff
@@ -111,43 +122,169 @@ private long getMaxReplicatedStreamsTail() {
 
-        // No data to send on the Active, so no replication remaining
+        // No data to send on the active, so no replication remaining
         if (maxReplicatedStreamTail == Address.NON_ADDRESS) {
+            log.debug("No data to replicate, replication complete.");
             return NO_REPLICATION_REMAINING_ENTRIES;
         }
 
-        // If doing a snapshot sync and nothing has been acked, all replication is remaining.  So set ack=0
-        if (ackedTimestamp == Address.NON_ADDRESS &&
-                lastSyncType == LogReplicationMetadata.ReplicationStatusVal.SyncType.SNAPSHOT) {
-            ackedTimestamp = 0;
+        if (lastSyncType.equals(LogReplicationMetadata.ReplicationStatusVal.SyncType.SNAPSHOT)) {
+
+            // If during snapshot sync nothing has been ack'ed, all replication is remaining
+            if (ackedTimestamp == Address.NON_ADDRESS) {
+                ackedTimestamp = 0;
+            }
+
+            // In Snapshot Sync
+            // Simply subtract the ack'ed timestamp from the global log tail from the time the snapshot sync started.
+            // Note that this is not accurate because the global log tail does not accurately represent the remaining entries
+            // for replicated streams.
+            // When snapshot sync is ongoing, there may be delta updates also. Add those new entries by querying the address maps
+            return ((baseSnapshotTimestamp - ackedTimestamp) +
+                    getTxStreamTotalEntries(baseSnapshotTimestamp, txStreamTail));
         }
 
-        // When in LogEntry Sync, no CP and trim has taken place so the remaining entries can be queried using the
-        // global tail and address maps of the replicated streams
-        if (lastSyncType == LogReplicationMetadata.ReplicationStatusVal.SyncType.LOG_ENTRY) {
-            return calculateRemainingEntriesForLogEntrySync(maxReplicatedStreamTail, ackedTimestamp);
+        // In Log Entry Sync
+        return calculateRemainingEntriesIncrementalUpdates(ackedTimestamp, txStreamTail, currentTxStreamProcessedTs);
+    }
+
+    private long getTxStreamTail(Map<UUID, Long> tailMap) {
+        if (tailMap.containsKey(TRANSACTION_STREAM_ID)) {
+            return tailMap.get(TRANSACTION_STREAM_ID);
         }
 
-        // In Snapshot Sync
-        // Simply subtract the ackedTimestamp from the global log tail from the time the snapshot sync started.
-        // Note that this is not accurate because the global log tail does not accurately represent the remaining entries
-        // for replicated streams.
-        // When snapshot sync is ongoing, there may be delta updates also.  Add those new entries by querying the address maps
-        return ((baseSnapshotTimestamp - ackedTimestamp) +
-            calculateRemainingEntriesForLogEntrySync(maxReplicatedStreamTail, baseSnapshotTimestamp));
+        log.warn("Tx Stream tail not present in sequencer, id={}", TRANSACTION_STREAM_ID);
```

Comment:
I don't think it can ever happen... but I was just being safe in case the get on that map does not find it... we can assume that stream has not been written so there is nothing remaining.